File: 5_kyu/Directions_Reduction.kt
package solution

object DirReduction {

    // Map each direction to the opposite direction
    val opposite =
        hashMapOf("NORTH" to "SOUTH", "SOUTH" to "NORTH", "EAST" to "WEST", "WEST" to "EAST")

    // Returns reduced direction array skipping all opposite direction pairs in the array.
    fun dirReduc(arr: Array<String>): Array<String> {
        var accu = mutableListOf<String>()
        arr.onEach {
            if (accu.isEmpty() || !(accu.last() == opposite[it])) accu.add(it)
            else accu.removeLast()
        }
        return accu.toTypedArray()
    }
}

File: 5_kyu/Directions_Reduction.r
# Returns reduced direction array skipping all opposite direction pairs.
dirReduc <- function(arr) {
  result <- list()
  for( it in arr )
    result <-
      if( length(result) == 0 || result[length(result)] != opposite_direction[[it]] )
        append(result, it)
      else
        result[-length(result)]
  unlist(result)
}

# Mapping from direction to opposite direction
# Values are the oppposite direction to the names
opposite_direction <- c(
  "NORTH"="SOUTH",
  "SOUTH"="NORTH",
  "WEST"="EAST",
  "EAST"="WEST" )
